replit/prybar
# Fix __main__.__file__ issue when imported from libs; unify python source code

# Why?

CMU-graphics has an [issue with prybar](https://linear.app/replit/issue/DX-163/%E2%9C%88%EF%B8%8Fpython-cmu-graphics-template-doesnt-run). It happens because a library tries to find the directory name of the entrypoint script by: `import __main__` and then accessing `__main__.__file__` which gives the full absolute path of the file. This works:

* in python 3.10 and 3.11
* not in python 3.8 (only the file name, not full path)
* not in any prybar version (only the file name, not full path)

We are going to fix it in prybar 3.10 and 3.11. Funnily this bug exists in python 3.8 as well, but that's not our problem.

## Changes

1. Added code to resolve absolute path of the entrypoint file.
2. #ifdefs: your favorite
3. re-orged the python code. Now it's a 3-in-1. Only one `languages/python3` subdirectory that is dry.

## Test

1. create `main.py`:
```py
import mylib
```
2. create `mylib.py`
```
import __main__
print(__main__.__file__)
```
3. `prybar-python3{x} main.py` you should see the full path of `main.py` for x = 310 and 311, but not 38.
